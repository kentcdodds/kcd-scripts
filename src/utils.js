const fs = require('fs')
const path = require('path')
const rimraf = require('rimraf')
const mkdirp = require('mkdirp')
const arrify = require('arrify')
const has = require('lodash.has')
const readPkgUp = require('read-pkg-up')
const which = require('which')
const {cosmiconfigSync} = require('cosmiconfig')

const {packageJson: pkg, path: pkgPath} = readPkgUp.sync({
  cwd: fs.realpathSync(process.cwd()),
})
const appDirectory = path.dirname(pkgPath)

function resolveKcdScripts() {
  if (pkg.name === 'kcd-scripts') {
    return require.resolve('./').replace(process.cwd(), '.')
  }
  return resolveBin('kcd-scripts')
}

// eslint-disable-next-line complexity
function resolveBin(modName, {executable = modName, cwd = process.cwd()} = {}) {
  let pathFromWhich
  try {
    pathFromWhich = fs.realpathSync(which.sync(executable))
    if (pathFromWhich && pathFromWhich.includes('.CMD')) return pathFromWhich
  } catch (_error) {
    // ignore _error
  }
  try {
    const modPkgPath = require.resolve(`${modName}/package.json`)
    const modPkgDir = path.dirname(modPkgPath)
    const {bin} = require(modPkgPath)
    const binPath = typeof bin === 'string' ? bin : bin[executable]
    const fullPathToBin = path.join(modPkgDir, binPath)
    if (fullPathToBin === pathFromWhich) {
      return executable
    }
    return fullPathToBin.replace(cwd, '.')
  } catch (error) {
    if (pathFromWhich) {
      return executable
    }
    throw error
  }
}

const fromRoot = (...p) => path.join(appDirectory, ...p)
const hasFile = (...p) => fs.existsSync(fromRoot(...p))
const ifFile = (files, t, f) =>
  arrify(files).some(file => hasFile(file)) ? t : f

const hasPkgProp = props => arrify(props).some(prop => has(pkg, prop))

const hasPkgSubProp = pkgProp => props =>
  hasPkgProp(arrify(props).map(p => `${pkgProp}.${p}`))

const ifPkgSubProp = pkgProp => (props, t, f) =>
  hasPkgSubProp(pkgProp)(props) ? t : f

const hasScript = hasPkgSubProp('scripts')
const hasPeerDep = hasPkgSubProp('peerDependencies')
const hasDep = hasPkgSubProp('dependencies')
const hasDevDep = hasPkgSubProp('devDependencies')
const hasAnyDep = args => [hasDep, hasDevDep, hasPeerDep].some(fn => fn(args))

const ifPeerDep = ifPkgSubProp('peerDependencies')
const ifDep = ifPkgSubProp('dependencies')
const ifDevDep = ifPkgSubProp('devDependencies')
const ifAnyDep = (deps, t, f) => (hasAnyDep(arrify(deps)) ? t : f)
const ifScript = ifPkgSubProp('scripts')

const hasTypescript = hasAnyDep('typescript') && hasFile('tsconfig.json')
const ifTypescript = (t, f) => (hasTypescript ? t : f)

function parseEnv(name, def) {
  if (envIsSet(name)) {
    try {
      return JSON.parse(process.env[name])
    } catch (err) {
      return process.env[name]
    }
  }
  return def
}

function envIsSet(name) {
  return (
    process.env.hasOwnProperty(name) &&
    process.env[name] &&
    process.env[name] !== 'undefined'
  )
}

function getConcurrentlyArgs(scripts, {killOthers = true} = {}) {
  const colors = [
    'bgBlue',
    'bgGreen',
    'bgMagenta',
    'bgCyan',
    'bgWhite',
    'bgRed',
    'bgBlack',
    'bgYellow',
  ]
  scripts = Object.entries(scripts).reduce((all, [name, script]) => {
    if (script) {
      all[name] = script
    }
    return all
  }, {})
  const prefixColors = Object.keys(scripts)
    .reduce(
      (pColors, _s, i) =>
        pColors.concat([`${colors[i % colors.length]}.bold.reset`]),
      [],
    )
    .join(',')

  // prettier-ignore
  return [
    killOthers ? '--kill-others-on-fail' : null,
    '--prefix', '[{name}]',
    '--names', Object.keys(scripts).join(','),
    '--prefix-colors', prefixColors,
    ...Object.values(scripts).map(s => JSON.stringify(s)), // stringify escapes quotes âœ¨
  ].filter(Boolean)
}

function uniq(arr) {
  return Array.from(new Set(arr))
}

function writeExtraEntry(name, {cjs, esm}, clean = true) {
  if (clean) {
    rimraf.sync(fromRoot(name))
  }
  mkdirp.sync(fromRoot(name))

  const pkgJson = fromRoot(`${name}/package.json`)
  const entryDir = fromRoot(name)

  fs.writeFileSync(
    pkgJson,
    JSON.stringify(
      {
        main: path.relative(entryDir, cjs),
        'jsnext:main': path.relative(entryDir, esm),
        module: path.relative(entryDir, esm),
      },
      null,
      2,
    ),
  )
}

function hasLocalConfig(moduleName, searchOptions = {}) {
  const explorerSync = cosmiconfigSync(moduleName, searchOptions)
  const result = explorerSync.search(pkgPath)

  return result !== null
}

module.exports = {
  appDirectory,
  fromRoot,
  getConcurrentlyArgs,
  hasFile,
  hasLocalConfig,
  hasPkgProp,
  hasScript,
  hasDep,
  ifAnyDep,
  ifDep,
  ifDevDep,
  ifFile,
  ifPeerDep,
  ifScript,
  hasTypescript,
  ifTypescript,
  parseEnv,
  pkg,
  resolveBin,
  resolveKcdScripts,
  uniq,
  writeExtraEntry,
}
