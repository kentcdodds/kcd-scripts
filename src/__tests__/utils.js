jest.mock('read-pkg-up', () => ({
  sync: jest.fn(() => ({pkg: {}, path: '/blah/package.json'})),
}))
jest.mock('which', () => ({sync: jest.fn(() => {})}))

let whichSyncMock, readPkgUpSyncMock

beforeEach(() => {
  jest.resetModules()
  whichSyncMock = require('which').sync
  readPkgUpSyncMock = require('read-pkg-up').sync
})

test('pkg is the package.json', () => {
  const myPkg = {name: 'blah'}
  mockPkg({pkg: myPkg})
  expect(require('../utils').pkg).toBe(myPkg)
})

test('appDirectory is the dirname to the package.json', () => {
  const pkgPath = '/some/path/to'
  mockPkg({path: `${pkgPath}/package.json`})
  expect(require('../utils').appDirectory).toBe(pkgPath)
})

test('resolveKcdScripts resolves to src/index.js when in the kcd-scripts package', () => {
  mockPkg({pkg: {name: 'kcd-scripts'}})
  expect(require('../utils').resolveKcdScripts()).toBe(
    require.resolve('../').replace(process.cwd(), '.'),
  )
})

test('resolveKcdScripts resolves to kcd-scripts if not in the kcd-scripts package', () => {
  mockPkg({pkg: {name: 'not-kcd-scripts'}})
  whichSyncMock.mockImplementationOnce(() => require.resolve('../'))
  expect(require('../utils').resolveKcdScripts()).toBe('kcd-scripts')
})

test(`resolveBin resolves to the full path when it's not in $PATH`, () => {
  expect(require('../utils').resolveBin('cross-env')).toBe(
    require.resolve('cross-env/dist/bin/cross-env').replace(process.cwd(), '.'),
  )
})

test(`resolveBin resolves to the binary if it's in $PATH`, () => {
  whichSyncMock.mockImplementationOnce(() =>
    require.resolve('cross-env/dist/bin/cross-env').replace(process.cwd(), '.'),
  )
  expect(require('../utils').resolveBin('cross-env')).toBe('cross-env')
  expect(whichSyncMock).toHaveBeenCalledTimes(1)
  expect(whichSyncMock).toHaveBeenCalledWith('cross-env')
})

describe('for windows', () => {
  let realpathSync

  beforeEach(() => {
    jest.doMock('fs', () => ({realpathSync: jest.fn()}))
    realpathSync = require('fs').realpathSync
  })
  afterEach(() => {
    jest.unmock('fs')
  })

  test('resolveBin resolves to .bin path when which returns a windows-style cmd', () => {
    const fullBinPath = '\\project\\node_modules\\.bin\\concurrently.CMD'
    realpathSync.mockImplementation(() => fullBinPath)
    expect(require('../utils').resolveBin('concurrently')).toBe(fullBinPath)
    expect(realpathSync).toHaveBeenCalledTimes(2)
  })
})

test('getConcurrentlyArgs gives good args to pass to concurrently', () => {
  expect(
    require('../utils').getConcurrentlyArgs({
      build: 'echo build',
      lint: 'echo lint',
      test: 'echo test',
      validate: 'echo validate',
      a: 'echo a',
      b: 'echo b',
      c: 'echo c',
      d: 'echo d',
      e: 'echo e',
      f: 'echo f',
      g: 'echo g',
      h: 'echo h',
      i: 'echo i',
      j: 'echo j',
    }),
  ).toMatchSnapshot()
})

test('parseEnv parses the existing environment variable', () => {
  const globals = {react: 'React', 'prop-types': 'PropTypes'}
  process.env.BUILD_GLOBALS = JSON.stringify(globals)
  expect(require('../utils').parseEnv('BUILD_GLOBALS')).toEqual(globals)
  delete process.env.BUILD_GLOBALS
})

test(`parseEnv returns the default if the environment variable doesn't exist`, () => {
  const defaultVal = {hello: 'world'}
  expect(require('../utils').parseEnv('DOES_NOT_EXIST', defaultVal)).toBe(
    defaultVal,
  )
})

test('ifAnyDep returns the true argument if true and false argument if false', () => {
  mockPkg({pkg: {peerDependencies: {react: '*'}}})
  const t = {a: 'b'}
  const f = {c: 'd'}
  expect(require('../utils').ifAnyDep('react', t, f)).toBe(t)
  expect(require('../utils').ifAnyDep('preact', t, f)).toBe(f)
})

test('ifAnyDep works with arrays of dependencies', () => {
  mockPkg({pkg: {peerDependencies: {react: '*'}}})
  const t = {a: 'b'}
  const f = {c: 'd'}
  expect(require('../utils').ifAnyDep(['preact', 'react'], t, f)).toBe(t)
  expect(require('../utils').ifAnyDep(['preact', 'webpack'], t, f)).toBe(f)
})

test('ifScript returns the true argument if true and the false argument if false', () => {
  mockPkg({pkg: {scripts: {build: 'echo build'}}})
  const t = {e: 'f'}
  const f = {g: 'h'}
  expect(require('../utils').ifScript('build', t, f)).toBe(t)
  expect(require('../utils').ifScript('lint', t, f)).toBe(f)
})

test('ifFile returns the true argument if true and the false argument if false', () => {
  mockPkg({path: require.resolve('../../package.json')})
  const t = {e: 'f'}
  const f = {g: 'h'}
  expect(require('../utils').ifFile('package.json', t, f)).toBe(t)
  expect(require('../utils').ifFile('does-not-exist.blah', t, f)).toBe(f)
})

function mockPkg({pkg = {}, path = '/blah/package.json'}) {
  readPkgUpSyncMock.mockImplementationOnce(() => ({pkg, path}))
}
