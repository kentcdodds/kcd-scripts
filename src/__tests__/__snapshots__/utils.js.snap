// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getConcurrentlyArgs gives good args to pass to concurrently 1`] = `
Array [
  --kill-others-on-fail,
  --prefix,
  [{name}],
  --names,
  build,lint,test,validate,a,b,c,d,e,f,g,h,i,j,
  --prefix-colors,
  bgBlue.bold.white,bgGreen.bold.white,bgMagenta.bold.white,bgCyan.bold.white,bgWhite.bold.white,bgRed.bold.white,bgBlack.bold.white,bgYellow.bold.white,bgBlue.bold.white,bgGreen.bold.white,bgMagenta.bold.white,bgCyan.bold.white,bgWhite.bold.white,bgRed.bold.white,
  "echo build",
  "echo lint",
  "echo test",
  "echo validate",
  "echo a",
  "echo b",
  "echo c",
  "echo d",
  "echo e",
  "echo f",
  "echo g",
  "echo h",
  "echo i",
  "echo j",
]
`;
